### PR TITLE
Support sm4 encryption algorithm

### LINKED ISSUE #5714: Encryption: Support SM4 Cryptographic Hash Algorithm
## Enhancement

Related tidb issue: https://github.com/pingcap/tidb/issues/36192
Related tikv issue: https://github.com/tikv/tikv/issues/13041

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close #5714 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Supports Chinese national standard encryption algorithm SM4 for encryption at rest.
```

### CODE REVIEW

File: release-centos7-llvm/dockerfiles/tiflash-llvm-base-aarch64 [hunk 1]
```diff
@@ -22,6 +22,7 @@ ENV PATH="/opt/cmake/bin:/usr/local/bin/:${PATH}:/usr/local/go/bin:/root/.cargo/
     LIBRARY_PATH="/usr/local/lib/aarch64-unknown-linux-gnu/:${LIBRARY_PATH}" \
     LD_LIBRARY_PATH="/usr/local/lib/aarch64-unknown-linux-gnu/:${LD_LIBRARY_PATH}" \
     CPLUS_INCLUDE_PATH="/usr/local/include/aarch64-unknown-linux-gnu/c++/v1/:${CPLUS_INCLUDE_PATH}" \
+    OPENSSL_ROOT_DIR="/usr/local/opt/openssl" \
```
Comment (JaySon-Huang) [inline release-centos7-llvm/dockerfiles/tiflash-llvm-base-aarch64:25]: Add "openssl"/"boringssl" in `TiFlashBuildInfo::getEnabledFeatures` may be helpful for check whether it is built with openssl

Comment (lidezhu) [reply release-centos7-llvm/dockerfiles/tiflash-llvm-base-aarch64:25]: Added.

File: dbms/src/Common/TiFlashBuildInfo.cpp [hunk 2]
```diff
@@ -51,15 +53,25 @@ std::string getUTCBuildTime()
...
+            "jemalloc",
 #elif USE_MIMALLOC
-        "mimalloc",
+            "mimalloc",
+#endif
+
+// OpenSSL or BoringSSL
+#if USE_INTERNAL_SSL_LIBRARY
+            "boringssl",
+#elif OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
```
Comment (JaySon-Huang) [inline dbms/src/Common/TiFlashBuildInfo.cpp:71]: where is `OPENSSL_NO_SM4` defined

Comment (lidezhu) [reply dbms/src/Common/TiFlashBuildInfo.cpp]: It is defined in <openssl/opensslconf.h>. And I remember BoringSSL will define `OPENSSL_NO_SM4` so I simplify the logic here by avoiding use `USE_INTERNAL_SSL_LIBRARY`.

File: dbms/src/Encryption/AESCTRCipherStream.cpp [hunk 3]
```diff
@@ -39,6 +39,14 @@ size_t KeySize(EncryptionMethod method)
         return 24;
     case EncryptionMethod::Aes256Ctr:
         return 32;
+    case EncryptionMethod::SM4Ctr:
+#if OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException("Unsupported encryption method: " + std::to_string(static_cast<int>(method)),
+                                   Errors::Encryption::Internal);
+#else
+        // Openssl support SM4 after 1.1.1 release version.
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipherStream.cpp:47]:
```suggestion
        // OpenSSL support SM4 after 1.1.1 release version.
```

File: dbms/src/Encryption/AESCTRCipherStream.cpp [hunk 4]
```diff
@@ -89,13 +98,14 @@ void AESCTRCipherStream::cipher(uint64_t file_offset, char * data, size_t data_s
     ret = EVP_CIPHER_CTX_set_padding(ctx, 0);
     if (ret != 1)
     {
+        FreeCipherContext(ctx);
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipherStream.cpp:101]: Use `SCOPE_EXIT` to free the cipher context?

File: dbms/src/Encryption/AESCTRCipherStream.cpp [hunk 5]
```diff
@@ -189,6 +205,15 @@ BlockAccessCipherStreamPtr AESCTRCipherStream::createCipherStream(
     case EncryptionMethod::Aes256Ctr:
         cipher = EVP_aes_256_ctr();
         break;
+    case EncryptionMethod::SM4Ctr:
+#if OPENSSL_VERSION_NUMBER < 0x1010100fL || defined(OPENSSL_NO_SM4)
+        throw DB::TiFlashException("Unsupported encryption method: " + std::to_string(static_cast<int>(encryption_info_.method)),
+                                   Errors::Encryption::Internal);
+#else
+        // Openssl support SM4 after 1.1.1 release version.
+        cipher = EVP_sm4_ctr();
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/AESCTRCipherStream.cpp:214]: Should we call `EVP_CIPHER_free` if exception throw after, and in `~ AESCTRCipherStream`?

https://github.com/openssl/openssl/blob/7f4c65749f7e9e687d5aa38e9b3fb548705511b9/demos/cipher/ariacbc.c#L163

File: dbms/src/Encryption/AESCTRCipherStream.cpp (hunk 4 above)
Comment (lidezhu) [reply dbms/src/Encryption/AESCTRCipherStream.cpp]: Fixed.

File: dbms/src/Encryption/AESCTRCipherStream.cpp (hunk 5 above)
Comment (lidezhu) [reply dbms/src/Encryption/AESCTRCipherStream.cpp:203]: `EVP_CIPHER_free` should only be called on cipher which is get from `EVP_CIPHER_fetch` as shown in https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_fetch.html. 
And from the source code it is shown that `EVP_xxx_ctr` just return a pointer to an initialized static storage. So it don't need to be freed after use.

**Comment by lidezhu** [comment]
@JaySon-Huang @flowbehappy @CalvinNeo PTAL

File: dbms/src/Encryption/AESCTRCipherStream.cpp (hunk 5 above)
Comment (JaySon-Huang) [reply dbms/src/Encryption/AESCTRCipherStream.cpp:203]: We can add some comment about it

**Review by JaySon-Huang — APPROVED** [review]
LGTM with minor comment

File: dbms/src/Encryption/AESCTRCipherStream.cpp (hunk 5 above)
Comment (lidezhu) [reply dbms/src/Encryption/AESCTRCipherStream.cpp:203]: Added in https://github.com/pingcap/tiflash/pull/5655/files/#diff-bcc8e47c4230c713d3018b3907a18e69cd02c7307396510e87618c41c2986033R184.

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**